src/core/state.go

```diff
@@ -491,17 +497,72 @@ func (state *BuildState) LogResult(result *BuildResult) {
 	}
 }
 
+// forwardResults runs indefinitely, forwarding results from the internal
+// channel to the external one. On the way it checks if we need to do
+// cycle detection.
+func (state *BuildState) forwardResults() {
+	defer func() {
+		if r := recover(); r != nil {
+			// Ensure we don't get a "send on closed channel" when the
+			// outward results channel is closed.
+			log.Debug("%s", r)
+		}
+	}()
+	activeTargets := map[*BuildTarget]struct{}{}
+	// Persist this one timer throughout so we don't generate bazillions of them.
+	t := time.NewTimer(cycleCheckDuration)
+	t.Stop()
+	var result *BuildResult
+	for {
+		if len(activeTargets) == 0 {
+			t.Reset(cycleCheckDuration)
+			select {
+			case result = <-state.progress.internalResults:
+				// This has to be properly managed to prevent hangs.
+				if !t.Stop() {
+					<-t.C
+				}
+			case <-t.C:
+				go state.checkForCycles()
+				// Still need to get a result!
+				result = <-state.progress.internalResults
```

Comment:
It will only continue if something does manage to progress - so for us to start overloading the system we'd have to have events emitted more slowly than 1 every 2 seconds, but also fast enough to re-trigger this quickly. That's _theoretically_ possible but seems so unlikely to happen that way in practice that I wasn't really worrying about it.

Reply:
Ah I see, I had a minor misunderstanding where I thought we could trigger these on builds finishing if we had no newly active targets coming in. Looks like we wait all active targets to finish before starting the timer though.